MosaicResidentInformationApi/V1/Gateways/MosaicGateway.cs

```diff
@@ -15,35 +20,114 @@ public class MosaicGateway : IMosaicGateway
 
         public MosaicGateway(MosaicContext mosaicContext)
         {
+            Console.WriteLine("In gateway constructor");
             _mosaicContext = mosaicContext;
         }
 
         public List<ResidentInformation> GetAllResidents(int cursor, int limit, string firstname = null,
             string lastname = null, string postcode = null, string address = null)
         {
-            var addressesFilteredByPostcode = _mosaicContext.Addresses
-                .Include(p => p.Person)
-                .Where(a => string.IsNullOrEmpty(address) || a.AddressLines.ToLower().Replace(" ", "").Contains(StripString(address)))
-                .Where(a => string.IsNullOrEmpty(postcode) || a.PostCode.ToLower().Replace(" ", "").Equals(StripString(postcode)))
-                .Where(a => string.IsNullOrEmpty(firstname) || a.Person.FirstName.ToLower().Replace(" ", "").Contains(StripString(firstname)))
-                .Where(a => string.IsNullOrEmpty(lastname) || a.Person.LastName.ToLower().Replace(" ", "").Contains(StripString(lastname)))
-                .Where(a => a.Person.Id > cursor)
+            var addressSearchPattern = GetSearchPattern(address);
+            var postcodeSearchPattern = GetSearchPattern(postcode);
+
+            var queryByAddress = postcode != null || address != null;
+
+            var peopleIds = queryByAddress
+                ? PeopleIdsForAddressQuery(cursor, limit, firstname, lastname, postcode, address)
+                : PeopleIds(cursor, limit, firstname, lastname);
+
+            var dbRecords = _mosaicContext.Persons
+                .Where(p => peopleIds.Contains(p.Id))
+                .Select(p => new
+                {
+                    Person = p,
+                    Addresses = _mosaicContext
+                        .Addresses
+                        .Where(add => add.PersonId == p.Id)
+                        .Where(add =>
+                            string.IsNullOrEmpty(address) || EF.Functions.ILike(add.AddressLines.Replace(" ", ""), addressSearchPattern))
+                        .Where(add =>
+                            string.IsNullOrEmpty(postcode) || EF.Functions.ILike(add.PostCode.Replace(" ", ""), postcodeSearchPattern))
+                        .Distinct()
+                        .ToList(),
+                    TelephoneNumbers = _mosaicContext.TelephoneNumbers.Where(n => n.PersonId == p.Id).Distinct().ToList()
+                }).ToList();
+
+            return dbRecords.Select(x => new ResidentInformation
+            {
```

Comment:
I think we can use the method `MapPersonAndAddressesToResidentInformation ` here

Reply:
Yep. Nice! 👍

Reply:
Added in d454b7a